spokestack/asr/google/speech_recognizer.py

```diff
@@ -0,0 +1,119 @@
+        **kwargs (optional): additional keyword arguments
+    """
+
+    def __init__(
+        self, language, credentials=None, sample_rate=16000, **kwargs,
+    ) -> None:
+        if credentials:
+            if isinstance(credentials, str):
+                credentials = service_account.Credentials.from_service_account_file(
+                    credentials
+                )
+            else:
+                credentials = service_account.Credentials.from_service_account_info(
+                    credentials
+                )
+
+        self._client = speech.SpeechClient(credentials=credentials)
+        self._config = speech.StreamingRecognitionConfig(
+            config=speech.RecognitionConfig(
+                encoding=speech.RecognitionConfig.AudioEncoding.LINEAR16,
+                sample_rate_hertz=sample_rate,
+                language_code=language,
+                enable_automatic_punctuation=True,
+            ),
+            interim_results=True,
+        )
+        self._sample_rate = sample_rate
+        self._language = language
+        self._queue: Queue = Queue()
+        self._thread: Any = None
+
+    def __call__(self, context: SpeechContext, frame: np.ndarray) -> None:
+        """Main entry point.
+
+        Args:
+            context (SpeechContext): current state of the speech pipeline
+            frame (np.ndarray): numpy array of PCM-16 audio.
+
+        Returns: None
+
+        """
+        if self._thread is None and context.is_active:
+            self._begin(context)
+        if self._thread is not None and not context.is_active:
+            self._commit()
+        if context.is_active:
+            self._send(frame)
+
+    def _begin(self, context) -> None:
+        self._thread = Thread(target=self._receive, args=(context,),)
+        self._thread.start()
+
+    def _receive(self, context):
+
+        for response in self._client.streaming_recognize(self._config, self._drain()):
+            for result in response.results[:1]:
+                for alternative in result.alternatives[:1]:
+                    context.transcript = alternative.transcript
+                    context.confidence = alternative.confidence
+
+                if result.is_final:
+                    if context.transcript:
+                        context.event("recognize")
+                        _LOG.debug("recognize event")
+                else:
+                    context.event("partial_recognize")
+                    _LOG.debug("partial recognize event")
+
+    def _drain(self):
+        while data := self._queue.get():
+            yield data
+
+    def _commit(self) -> None:
+        self._queue.put(None)
+        self._thread.join()
+        self._thread = None
+
+    def _send(self, frame) -> None:
+        self._queue.put(speech.StreamingRecognizeRequest(audio_content=frame.tobytes()))
+
+    def reset(self) -> None:
+        """ resets recognizer """
+        self._queue.put(None)
```

Comment:
adding a value to the queue and then emptying it does nothing
also, there is no need to write a value to the queue if the thread is not attached